pkg/controller/installation/products/fuse/reconciler.go

```diff
@@ -0,0 +1,208 @@
+	Config        *config.Fuse
+	ConfigManager config.ConfigReadWriter
+	mpm           marketplace.MarketplaceInterface
+	logger        *logrus.Entry
+	nsReconciler  *resources.NamespaceReconciler
+}
+
+func NewReconciler(client pkgclient.Client, rc *rest.Config, coreClient *kubernetes.Clientset, configManager config.ConfigReadWriter, instance *v1alpha1.Installation, logger *logrus.Entry) (*Reconciler, error) {
+	fuseConfig, err := configManager.ReadFuse()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not retrieve keycloak codeReadyConfig")
+	}
+
+	if fuseConfig.GetNamespace() == "" {
+		fuseConfig.SetNamespace(instance.Spec.NamespacePrefix + defaultInstallationNamespace)
+	}
+	if err = fuseConfig.Validate(); err != nil {
+		return nil, errors.Wrap(err, "keycloak codeReadyConfig is not valid")
+	}
+
+	mpm := marketplace.NewManager(client, rc)
+	nsr := resources.NewNamespaceReconciler(client, logger)
+	return &Reconciler{client: client,
+		coreClient:    coreClient,
+		ConfigManager: configManager,
+		Config:        fuseConfig,
+		mpm:           mpm,
+		logger:        logger,
+		nsReconciler:  nsr,
+	}, nil
+}
+
+func (r *Reconciler) Reconcile(inst *v1alpha1.Installation) (v1alpha1.StatusPhase, error) {
+	ctx := context.TODO()
+
+	reconciledPhase, err := r.reconcileNamespace(ctx, inst)
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, " failed to reconcile namespace for fuse ")
+	}
+
+	reconciledPhase, err = r.reconcileSubscription(ctx)
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, " failed to reconcile subscription for fuse ")
+	}
+
+	reconciledPhase, err = r.handleAwaitingOperator(ctx)
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, " failed to reconcile subscription for fuse ")
+	}
+	reconciledPhase, err = r.reconcileCustomResource(ctx, inst)
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, " failed to reconcile fuse custom resource ")
+	}
+
+	logrus.Info("End of reconcile Phase : ", reconciledPhase)
+
+	return v1alpha1.StatusPhase(v1alpha1.PhaseCompleted), nil
+}
+
+func (r *Reconciler) reconcileNamespace(ctx context.Context, inst *v1alpha1.Installation) (v1alpha1.StatusPhase, error) {
+	if !inst.Spec.CreateNamespaces {
+		r.logger.Infof("skipping creating namespace: %s", r.Config.GetNamespace())
+		return v1alpha1.PhaseCreatingSubscription, nil
+	}
+	ns := &v1.Namespace{
+		ObjectMeta: v12.ObjectMeta{
+			Name: r.Config.GetNamespace(),
+		},
+	}
+	ns, err := r.nsReconciler.Reconcile(ctx, ns, inst)
+	if err != nil {
+		return v1alpha1.PhaseFailed, errors.Wrap(err, "failed to reconcile fuse namespace "+r.Config.GetNamespace())
+	}
+	if ns.Status.Phase == v1.NamespaceTerminating {
+		r.logger.Debugf("namespace %s is terminating, maintaining phase to try again on next reconcile", r.Config.GetNamespace())
+		return v1alpha1.PhaseAwaitingNS, nil
+	}
+	if ns.Status.Phase != v1.NamespaceActive {
+		return v1alpha1.PhaseAwaitingNS, nil
```

Comment:
Just a question to understand how this works, does returning here not cause it to go into the next phase?